src/Lecture4.hs

```diff
@@ -184,7 +218,17 @@ instance for the 'Stats' type itself.
 -}
 
 instance Semigroup Stats where
-
+   (<>) s1 s2 = Stats
+                { statsTotalPositions = statsTotalPositions s1 <> statsTotalPositions s2
+                , statsTotalSum       = statsTotalSum s1 <> statsTotalSum s2
+                , statsAbsoluteMax    = statsAbsoluteMax s1 <> statsAbsoluteMax s2
+                , statsAbsoluteMin    = statsAbsoluteMin s1 <> statsAbsoluteMin s2
+                , statsSellMax        = statsSellMax s1 <> statsSellMax s2
+                , statsSellMin        = statsSellMin s1 <> statsSellMin s2
+                , statsBuyMax         = statsBuyMax s1 <> statsBuyMax s2
+                , statsBuyMin         = statsBuyMin s1 <> statsBuyMin s2
+                , statsLongest        = statsLongest s1 <> statsLongest s2
+                }
```

Comment:
This is one of two fixes needed to implement the optimal streaming and lazy solution 🦥 🌊 

The reason for this change is that the `Just` constructor of `Maybe` is not strict and the `Semigroup` instance for `Maybe` doesn't force values inside `Just`.

```haskell
instance Semigroup a => Semigroup (Maybe a) where
    Nothing <> b       = b
    a       <> Nothing = a
    Just a  <> Just b  = Just (a <> b)
```

So you get the accumulation of functions `max` and `min` inside these constructors instead of a single value on each step. You can fix this by creating a custom "forcing" function that asks GHC to evaluate the value of each `Maybe` right now.

```suggestion
                , statsAbsoluteMin    = statsAbsoluteMin s1 <> statsAbsoluteMin s2
                , statsSellMax        = forceMaybe (statsSellMax s1 <> statsSellMax s2)
                , statsSellMin        = forceMaybe (statsSellMin s1 <> statsSellMin s2)
                , statsBuyMax         = forceMaybe (statsBuyMax s1 <> statsBuyMax s2)
                , statsBuyMin         = forceMaybe (statsBuyMin s1 <> statsBuyMin s2)
                , statsLongest        = statsLongest s1 <> statsLongest s2
                }
  where
    forceMaybe :: Maybe a -> Maybe a
    forceMaybe m = case m of
        Nothing -> Nothing
        Just !x -> Just x
```